### DISCUSSION TITLE
Mocking difficulties

### DISCUSSION THREAD

**Discussion by lowki** [prisma/prisma#19048] [Q&A]
Hi.
I have a monorepo stack with prisma and vitest and can't figure out to mock my client.
I'm following this article https://www.prisma.io/blog/testing-series-1-8eRB5p0Y8o to get help but without success.

Here what I have so far.

Prisma client export and function calling it
```ts
// packages/core/db/index.ts
// singleton client export

import { Prisma, PrismaClient } from './prisma/client';
const dbClient = new PrismaClient();
export default dbClient;
```

```ts
// packages/functions/myFunction.ts
import dbClient from '@/core/db';

export getUsers = async (name) => {
  const users = await dbClient.users.findMany({
    where: {
      name
    }
  });
return users;
};
```

Mock and test
```ts
// packages/core/db/__mocks__/db.ts
import { PrismaClient } from '../prisma/client';
import { beforeEach } from 'vitest';
import { mockDeep, mockReset } from 'vitest-mock-extended';

beforeEach(() => {
  mockReset(dbClient);
});

const dbClient = mockDeep<PrismaClient>();
export default dbClient;

```
```ts
// test/functions/myFunction.test.ts
import { it, vi, expect } from 'vitest';
import { getUsers } from '@/functions/myFunction';
import dbClient from '@/core/db/__mocks__/db';

vi.mock('@/core/db', () => {
  return {
    default: () => dbClient
  };
});

it('should return users matching name', async () => {
  const user = {id: 1, name: 'test'};
  dbClient.users.findMany.mockResolvedValue([user]);
  const result = await getUsers('test');
  expect(result).toMatchObject([user]);
});
```

Running this throw the following error, telling me, the mocked client is empty.
```console
TypeError: Cannot read properties of undefined (reading 'findMany')
```

Any help would be appreciated.

**Comment by ludralph** [comment]
Hi @lowki 👋 

From this line
> ```js
> // packages/core/db/index.ts
>// singleton client export
>
> import { Prisma, PrismaClient } from './prisma/client';
> const dbClient = new PrismaClient();
> export default dbClient;


Is the `import` not supposed to be from `@prisma/client`?

**Reply by lowki** [reply]
Well, not really, the actual Prisma client is at the same level.
```
db
├── __mocks__
│   └── db.ts
└── index.ts
    └── prisma
        ├── client
        └── index.js
```

**Reply by ludralph** [reply]
Could you please share a minimal reproduction so I can try this locally?

**Reply by lowki** [reply]
Sure, I made a repo to reproduce my issue, you should be able to use it like this:

```bash
git clone git@github.com:lowki/prisma-mocking-test.git
cd prisma-mocking-test
pnpm install
pnpm test
```

**Reply by chaba11** [reply]
Have you found a solution for this? I've been struggling with the same issue

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
